service/controller/clusterapi/v29/changedetection/tcnp.go

```diff
@@ -0,0 +1,88 @@
+package changedetection
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/key"
+)
+
+type TCNPConfig struct {
+	Logger micrologger.Logger
+}
+
+// TCNP is a detection service implementation deciding if a node pool should be
+// updated or scaled.
+type TCNP struct {
+	logger micrologger.Logger
+}
+
+func NewTCNP(config TCNPConfig) (*TCNP, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	t := &TCNP{
+		logger: config.Logger,
+	}
+
+	return t, nil
+}
+
+// ShouldScale determines whether the reconciled node pool should be scaled. A
+// node pool is only allowed to scale in the following cases.
+//
+//     The tenant cluster's scaling max changes.
+//     The tenant cluster's scaling min changes.
+//
+func (t *TCNP) ShouldScale(ctx context.Context, md v1alpha1.MachineDeployment) (bool, error) {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	if !cc.Status.TenantCluster.TCNP.ASG.IsEmpty() && cc.Status.TenantCluster.TCNP.ASG.MaxSize != key.MachineDeploymentScalingMax(md) {
```

Comment:
I'd propose to potentially break these down a bit for easier readability:
```go
asgExists := !cc.Status.TenantCluster.TCNP.ASG.IsEmpty()
asgMaxNumberOfNodesDiffer := cc.Status.TenantCluster.TCNP.ASG.MaxSize != key.MachineDeploymentScalingMax(md)
asgMinNumberOfNodesDiffer := cc.Status.TenantCluster.TCNP.ASG.MinSize != key.MachineDeploymentScalingMin(md)

if asgExists && asgMaxNumberOfNodesDiffer {
    // ....
}

if asgExists && asgMinNumberOfNodesDiffer {
    // ....
}
```

This way it's not necessary to list these in above function comment which might get forgotten to update when later adding more rules here. That kind of explicit variable naming makes the code basically the same as comment and therefore documents itself quite nicely.

Reply:
Good shout. Done that.